io-sim/src/Control/Monad/IOSimPOR/Internal.hs

```diff
@@ -0,0 +1,1574 @@
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+ 
+{-# OPTIONS_GHC -Wno-orphans            #-}
+-- incomplete uni patterns in 'schedule' (when interpreting 'StmTxCommitted')
+-- and 'reschedule'.
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns -Wno-unused-matches #-}
+
+module Control.Monad.IOSimPOR.Internal (
+  IOSim (..),
+  SimM,
+  runIOSim,
+  runSimTraceST,
+  traceM,
+  traceSTM,
+  STM,
+  STMSim,
+  SimSTM,
+  setCurrentTime,
+  unshareClock,
+  TimeoutException (..),
+  EventlogEvent (..),
+  EventlogMarker (..),
+  ThreadId,
+  ThreadLabel,
+  Labelled (..),
+  SimTrace,
+  TraceEvent,
+  liftST,
+  execReadTVar,
+  controlSimTraceST,
+  ScheduleControl(..),
+  ScheduleMod(..)
+  ) where
+  
+
+import           Prelude hiding (read)
+
+import           Data.Ord
+import           Data.Foldable (traverse_)
+import qualified Data.List as List
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.OrdPSQ (OrdPSQ)
+import qualified Data.OrdPSQ as PSQ
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Time (UTCTime (..), fromGregorian)
+
+import           Control.Exception (assert)
+import           Control.Monad (join)
+
+import           Control.Monad (when)
+import           Control.Monad.ST.Lazy
+import           Control.Monad.ST.Lazy.Unsafe (unsafeIOToST)
+import           Data.STRef.Lazy
+
+import           Control.Monad.Class.MonadSTM hiding (STM, TVar)
+import           Control.Monad.Class.MonadThrow as MonadThrow
+import           Control.Monad.Class.MonadTime
+import           Control.Monad.Class.MonadTimer
+
+import           Control.Monad.IOSim.Types
+import           Control.Monad.IOSimPOR.Timeout(unsafeTimeout)
+
+deriving instance Ord MaskingState
+
+--
+-- Simulation interpreter
+--
+
+data Thread s a = Thread {
+    threadId      :: !ThreadId,
+    threadControl :: !(ThreadControl s a),
+    threadBlocked :: !Bool,
+    threadDone    :: !Bool,
+    threadMasking :: !MaskingState,
+    -- other threads blocked in a ThrowTo to us because we are or were masked
+    threadThrowTo :: ![(SomeException, Labelled ThreadId, VectorClock)],
+    threadClockId :: !ClockId,
+    threadLabel   :: Maybe ThreadLabel,
+    threadNextTId :: !Int,
+    threadStep    :: !Int,
+    threadVClock  :: VectorClock,
+    threadEffect  :: Effect,  -- in the current step
+    threadRacy    :: !Bool
+  }
+  deriving Show
+
+-- We hide the type @b@ here, so it's useful to bundle these two parts
+-- together, rather than having Thread have an extential type, which
+-- makes record updates awkward.
+data ThreadControl s a where
+  ThreadControl :: SimA s b
+                -> ControlStack s b a
+                -> ThreadControl s a
+
+instance Show (ThreadControl s a) where
+  show _ = "..."
+
+data ControlStack s b a where
```

Comment:
Maybe they should go into an InternalTypes module, since Types does get imported into code that uses the library, and these aren't part of the interface.